cloudinit/sources/DataSourceNWCS.py

```diff
@@ -0,0 +1,162 @@
+        self.distro = distro
+        self.metadata = dict()
+        self.ds_cfg = util.mergemanydict(
+            [
+                util.get_cfg_by_path(sys_cfg, ["datasource", "NWCS"], {}),
+                BUILTIN_DS_CONFIG,
+            ]
+        )
+        self.metadata_address = self.ds_cfg["metadata_url"]
+        self.retries = self.ds_cfg.get("retries", MD_RETRIES)
+        self.timeout = self.ds_cfg.get("timeout", MD_TIMEOUT)
+        self.wait_retry = self.ds_cfg.get("wait_retry", MD_WAIT_RETRY)
+        self._network_config = sources.UNSET
+        self.dsmode = sources.DSMODE_NETWORK
+
+    def _get_data(self):
+        LOG.info("Detecting if machine is a NWCS instance")
+        on_nwcs = get_nwcs_data()
+
+        if not on_nwcs:
+            LOG.info("Machine is not a NWCS instance")
+            return False
+
+        LOG.info("Machine is a NWCS instance")
+
+        md = self.get_metadata()
+
+        if md is None:
+            raise Exception("failed to get metadata")
+
+        self.metadata_full = md
+
+        self.metadata["instance-id"]    = md["instance-id"]
+        self.metadata["public-keys"]    = md["public-keys"]
+        self.metadata["network"]        = md["network"]
+        self.metadata["local-hostname"] = md["hostname"]
+
+        self.userdata_raw = md.get("userdata", None)
+
+        self.vendordata_raw = md.get("vendordata", None)
+
+        return True
+
+    def get_metadata(self):
+        try:
+            LOG.info("Attempting to get metadata via DHCP")
+
+            with EphemeralDHCPv4(
+                iface=net.find_fallback_nic(),
+                connectivity_url_data={
+                    "url": BASE_URL_V1 + "/metadata/instance-id",
+                },
+            ):
+                return read_metadata(
+                    self.metadata_address,
+                    timeout=self.timeout,
+                    sec_between=self.wait_retry,
+                    retries=self.retries,
+                )
+            
+        except (
+            NoDHCPLeaseError,
+            subp.ProcessExecutionError,
+            RuntimeError,
+            exceptions.RequestException,
+        ) as e:
+            LOG.error("DHCP failure: %s", e)
+            raise
+
+    @property
+    def network_config(self):
+        LOG.debug("Attempting network configuration")
+
+        if self._network_config is None:
+            LOG.warning(
+                "Found None as cached _network_config, resetting to %s",
+                sources.UNSET,
+            )
+            self._network_config = sources.UNSET
+
+        if self._network_config != sources.UNSET:
+            return self._network_config
+
+        if not self.metadata["network"]["config"]:
+            raise Exception("Unable to get metadata from server")
+
+        # metadata sends interface names, but we dont want to use them
+        for i in self.metadata["network"]["config"]:
```

Comment:
This section feels a bit funny to me. Why aren't you trusting your metadata service? If you have multiple nics, how do you know which should be considered primary and which should be secondary?

Reply:
Are you referring to ignoring the interface names from the metadata service? I send the interface names to be more compatible with the API; but, I discard them in favor of what the underlying OS has for those MACs instead.